src/annotator/anchoring/test/xpath-util-test.js

```diff
@@ -0,0 +1,320 @@
+import $ from 'jquery';
+
+import {
+  findChild,
+  getTextNodes,
+  getLastTextNodeUpTo,
+  getFirstTextNodeNotBefore,
+  simpleXPathPure,
+  simpleXPathJQuery,
+} from '../xpath-util';
+
+describe('annotator/anchoring/xpath-util', () => {
+  describe('#findChild', () => {
+    let container;
+
+    beforeEach(() => {
+      container = document.createElement('div');
+      document.body.appendChild(container);
+    });
+
+    afterEach(() => {
+      container.remove();
+    });
+
+    it('throws an error when the parent has no children', () => {
+      assert.throws(() => {
+        findChild(container, 'fake', 0);
+      }, 'XPath error: node has no children!');
+    });
+
+    it('throws an error when the desired child is not found', () => {
+      container.innerHTML = `<div>text</div>`;
+      assert.throws(() => {
+        findChild(container, 'fake', 0);
+      }, 'XPath error: wanted child not found.');
+    });
+
+    describe('with children', () => {
+      beforeEach(() => {
+        container = document.createElement('div');
+        container.innerHTML = `text 1<span>text 2</span><span>text 3</span>text 4`;
+        document.body.appendChild(container);
+      });
+
+      afterEach(() => {
+        container.remove();
+      });
+
+      it('finds the desired text() node at index 1', () => {
+        assert.equal(findChild(container, 'text()', 1).nodeValue, 'text 1');
+        document.body.removeChild(container);
+      });
+
+      it('finds the desired text() node at index 2', () => {
+        assert.equal(findChild(container, 'text()', 2).nodeValue, 'text 4');
+      });
+
+      it('finds the desired span node at index 1', () => {
+        assert.equal(findChild(container, 'span', 1).innerText, 'text 2');
+      });
+
+      it('finds the desired span node at index 2', () => {
+        assert.equal(findChild(container, 'span', 2).innerText, 'text 3');
+      });
+    });
+  });
+
+  describe('#getTextNodes', () => {
+    let container;
+
+    beforeEach(() => {
+      container = document.createElement('div');
+      document.body.appendChild(container);
+    });
+
+    afterEach(() => {
+      container.remove();
+    });
+
+    it('finds basic text nodes', () => {
+      container.innerHTML = `<span>text 1</span><span>text 2</span>`;
+      const result = getTextNodes($(container));
+      assert.equal(result.length, 2);
+      assert.equal(result[0].nodeValue, 'text 1');
+      assert.equal(result[1].nodeValue, 'text 2');
+    });
+
+    it('finds basic text nodes and whitespace', () => {
+      container.innerHTML = `<span>text 1</span>
+        <span>text 2</span>`;
+      const result = getTextNodes($(container));
+      assert.equal(result.length, 3);
+    });
+
+    it('finds basic text nodes and whitespace but ignores comments', () => {
+      container.innerHTML = `<span>text 1</span>
+        <!--span>text 2</span-->`;
+      const result = getTextNodes($(container));
+      assert.equal(result.length, 2);
```

Comment:
This test only checks the first returned node. One pattern I quite like for checking the length and some aspect of the contents of each item of an array in tests is to do something like the following:

```js
// Helper
const nodeValues = nodes => nodes.map(n => n.nodeValue)

// In individual tests
const result = getTextNodes(...)
assert.deepEqual(nodeValues(result), ['text 1', 'text 2'])
```

Reply:
done!